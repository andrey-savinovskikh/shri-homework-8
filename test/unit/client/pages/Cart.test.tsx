import React from 'react';
import renderer from 'react-test-renderer';
import "@testing-library/jest-dom/extend-expect";
import {Provider} from "react-redux";
import {fireEvent, render, RenderResult, waitFor} from "@testing-library/react";
import {BrowserRouter} from "react-router-dom";

import {Cart} from "../../../../src/client/pages/Cart";
import {getStore} from "../../../utils/getStore";
import {CartState} from "../../../../src/common/types";
import {addToCart} from "../../../../src/client/store";
import {PRODUCTS} from "../../../utils/mockApi";

const getCart = (initialCartData: CartState = {}) => {
    const store = getStore(initialCartData);

    const cart = (
        <BrowserRouter>
            <Provider store={store}>
                <Cart/>
            </Provider>
        </BrowserRouter>
    )

    return {store, cart}
}

const getFormElements = (renderResult: RenderResult) => {
    const {getByLabelText, getByText} = renderResult;

    const name = getByLabelText("Name"),
        phone = getByLabelText("Phone"),
        address = getByLabelText("Address"),
        submit = getByText("Checkout");

    fireEvent.change(name, {target: {value: "Name" }});
    fireEvent.change(phone, {target: {value: "9999999999"}});
    fireEvent.change(address, {target: {value: "Address"}});

    return {name, phone, address, submit}
}

const cartData = {
    1: {
        name: "product-1",
        price: 1000,
        count: 5
    },
    2: {
        name: "product-2",
        price: 2000,
        count: 1
    }
};

describe('Страница Cart', () => {
    it('Корректный снапшот страницы с пустой корзиной', () => {
        const {cart} = getCart();
        const tree = renderer.create(cart).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Корректный снапшот страницы с заполненной корзиной', () => {
        const {cart} = getCart(cartData);
        const tree = renderer.create(cart).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("Если есть товары в корзине, отображается форма заказа", () => {
        const {cart} = getCart(cartData);
        const {container} = render(cart);

        expect(container.querySelector(".Form")).toBeInTheDocument();
    });

    it("Если корзина пуста, то форма заказа скрыта", () => {
        const {cart} = getCart();
        const {container} = render(cart);

        expect(container.querySelector(".Form")).not.toBeInTheDocument();
    });

    it('Работает очистка корзины', () => {
        const {cart} = getCart(cartData);
        const {container, getByText} = render(cart);

        const btn = container.querySelector(".Cart-Clear");
        fireEvent.click(btn);

        expect(getByText("Cart is empty", {exact: false})).toBeInTheDocument();
    });

    it("При успешной отправке формы заказа отображается оповещение", async () => {
        const {cart} = getCart(cartData);
        const renderResult = render(cart);
        const {container} = renderResult;

        const {submit} = getFormElements(renderResult);

        submit.click();

        await waitFor(() => {
            const successMessage = container.querySelector(".Cart-SuccessMessage");

            expect(successMessage.textContent).toContain("Well done!");
        });
    });

    it("Товар добавляется в корзину", async () => {
        const {cart, store} = getCart();

        store.dispatch(addToCart(PRODUCTS[0]));

        expect(store.getState().cart).toStrictEqual({
            [PRODUCTS[0].id]: {
                name: PRODUCTS[0].name,
                price: PRODUCTS[0].price,
                count: 1,
            }
        })

        store.dispatch(addToCart(PRODUCTS[0]));

        expect(store.getState().cart).toStrictEqual({
            [PRODUCTS[0].id]: {
                name: PRODUCTS[0].name,
                price: PRODUCTS[0].price,
                count: 2,
            }
        })
    });

    // it("При успешной отправке формы заказа проиходит очистка корзины", async () => {
    //     const {cart} = getCart(cartData);
    //     const renderResult = render(cart);
    //     const {getByText} = renderResult;
    //
    //     const {submit} = getFormElements(renderResult);
    //
    //     submit.click();
    //
    //     await waitFor(() => {
    //         expect(getByText("Cart is empty", {exact: false})).toBeInTheDocument();
    //     });
    // });
})