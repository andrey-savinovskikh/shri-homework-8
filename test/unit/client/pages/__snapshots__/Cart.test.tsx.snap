// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Страница Cart Корректный снапшот страницы с заполненной корзиной 1`] = `
<div
  className="Cart"
>
  <div
    className="row mb-4"
  >
    <div
      className="col"
    >
      <h1>
        Shopping cart
      </h1>
      <table
        className="Cart-Table table"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              #
            </th>
            <th
              scope="col"
            >
              Product
            </th>
            <th
              scope="col"
            >
              Price
            </th>
            <th
              scope="col"
            >
              Count
            </th>
            <th
              scope="col"
            >
              Total
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            data-testid="1"
          >
            <th
              className="Cart-Index"
              scope="row"
            >
              1
            </th>
            <td
              className="Cart-Name"
            >
              product-1
            </td>
            <td
              className="Cart-Price"
            >
              $
              1000
            </td>
            <td
              className="Cart-Count"
            >
              5
            </td>
            <td
              className="Cart-Total"
            >
              $
              5000
            </td>
          </tr>
          <tr
            data-testid="2"
          >
            <th
              className="Cart-Index"
              scope="row"
            >
              2
            </th>
            <td
              className="Cart-Name"
            >
              product-2
            </td>
            <td
              className="Cart-Price"
            >
              $
              2000
            </td>
            <td
              className="Cart-Count"
            >
              1
            </td>
            <td
              className="Cart-Total"
            >
              $
              2000
            </td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <td
              colSpan={4}
            >
              Order price:
            </td>
            <td
              className="Cart-OrderPrice"
            >
              $
              7000
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
  <div
    className="row mb-4"
  >
    <div
      className="col-6"
    >
      <button
        className="Cart-Clear btn btn-outline-secondary"
        onClick={[Function]}
      >
        Clear shopping cart
      </button>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-12 col-sm-8 col-md-6 col-lg-4"
    >
      <h2>
        Сheckout
      </h2>
      <div
        className="Form"
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="f-name"
          >
            Name
          </label>
          <input
            autoComplete="off"
            className="Form-Field Form-Field_type_name form-control"
            disabled={false}
            id="f-name"
            onChange={[Function]}
            type="text"
          />
          <div
            className="invalid-feedback"
          >
            Please provide your name
          </div>
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="f-phone"
          >
            Phone
          </label>
          <input
            className="Form-Field Form-Field_type_phone form-control"
            disabled={false}
            id="f-phone"
            onChange={[Function]}
            type="text"
          />
          <div
            className="invalid-feedback"
          >
            Please provide a valid phone
          </div>
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="f-address"
          >
            Address
          </label>
          <textarea
            className="Form-Field Form-Field_type_address form-control"
            disabled={false}
            id="f-address"
            onChange={[Function]}
            rows={3}
          />
          <div
            className="invalid-feedback"
          >
            Please provide a valid address
          </div>
        </div>
        <button
          className="Form-Submit btn btn-primary"
          disabled={false}
          onClick={[Function]}
        >
          Checkout
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Страница Cart Корректный снапшот страницы с пустой корзиной 1`] = `
<div
  className="Cart"
>
  <div
    className="row mb-4"
  >
    <div
      className="col"
    >
      <h1>
        Shopping cart
      </h1>
      Cart is empty. Please select products in the 
      <a
        href="/catalog"
        onClick={[Function]}
      >
        catalog
      </a>
      .
    </div>
  </div>
</div>
`;
